src/_Derivatives/subexpressions.jl

```diff
@@ -14,78 +14,100 @@ function list_subexpressions(nd::Vector{NodeData})
     return sort(collect(indices))
 end
 
-# Order the subexpressions which main_expressions depend on such that we can
-# run forward mode in this order.
-function order_subexpressions(
-    main_expressions::Vector{Vector{NodeData}},
+function _visit_nodes(
     subexpressions::Vector{Vector{NodeData}},
+    n::Int64,
+    individual::Tuple{Vector{Int64},Set{Int64}},
+    full::Union{Nothing,Tuple{Vector{Int64},Set{Int64}}},
 )
-    num_sub = length(subexpressions)
-    computed = falses(num_sub)
-    dependencies = Dict{Int,Vector{Int}}()
-    to_visit = collect(num_sub+1:num_sub+length(main_expressions))
-    # For each subexpression k, the indices of the main expressions that depend
-    # on k, possibly transitively.
-    depended_on_by = [Set{Int}() for i in 1:num_sub]
-
-    while !isempty(to_visit)
-        idx = pop!(to_visit)
-        if idx > num_sub
-            subexpr = list_subexpressions(main_expressions[idx-num_sub])
-        else
-            computed[idx] && continue
-            subexpr = list_subexpressions(subexpressions[idx])
-            computed[idx] = true
-        end
-        dependencies[idx] = subexpr
-        for k in subexpr
-            if idx > num_sub
-                push!(depended_on_by[k], idx - num_sub)
-            end
-            push!(to_visit, k)
-        end
+    # First, check if we have already visited node `n`. If so, bail from the
+    # recursion.
+    if n in individual[2]
+        return
     end
-
-    # Now order dependencies.
-    I = Int[]
-    J = Int[]
-    for (idx, subexpr) in dependencies
-        for k in subexpr
-            push!(I, idx)
-            push!(J, k)
+    # Else, recurse on each SUBEXPRESSION node in subexpression n.
+    for m in subexpressions[n]
+        if m.nodetype == SUBEXPRESSION
+            _visit_nodes(subexpressions, m.index, individual, full)
```

Comment:
I don't think we have the luxury of using recursion. Someone could easily make a chain of expressions that causes a stack overflow.